thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java

```diff
@@ -0,0 +1,320 @@
+            createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
+          }
+        } else {
+          datasetLastRefreshTime = dataFrame.getDoubles("timestamp").max().longValue();
+          if (isPartialData(datasetLastRefreshTime, info, datasetConfig)) {
+            if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+              createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
+            }
+          }
+        }
+      } else if (isPartialData(datasetLastRefreshTime, info, datasetConfig)) {
+        // Optimize for the common case - the common case is that the data availability events are arriving
+        // correctly and we need not re-fetch the data to double check.
+        if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+          createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
+        }
+      }
+    } catch (Exception e) {
+      LOG.error(String.format("Failed to run sla check on metric URN %s", me.getUrn()), e);
+    }
+  }
+
+  /**
+   * We say the data is missing we do not have data in the sla detection window.
+   * Or more specifically if the dataset watermark is below the sla detection window.
+   */
+  private boolean isMissingData(long datasetLastRefreshTime, DetectionPipelineTaskInfo info) {
+    return datasetLastRefreshTime < info.getStart();
+  }
+
+  /**
+   * We say the data is partial if we do not have all the data points in the sla detection window.
+   * Or more specifically if we have at least 1 data-point missing in the sla detection window.
+   *
+   * For example:
+   * Assume that the data is delayed and our current sla detection window is [1st Feb to 3rd Feb). During this scan,
+   * let's say data for 1st Feb arrives. Now, we have a situation where partial data is present. In other words, in the
+   * current sla detection window [1st to 3rd Feb) we have data for 1st Feb but data for 2nd Feb is missing. This is the
+   * partial data scenario.
+   */
+  private boolean isPartialData(long datasetLastRefreshTime, DetectionPipelineTaskInfo info, DatasetConfigDTO datasetConfig) {
+    long granularity = datasetConfig.bucketTimeGranularity().toMillis();
+    return (info.getEnd() - datasetLastRefreshTime) * 1.0 / granularity > 1;
+  }
+
+  /**
+   * Validates if the data is delayed or not based on the user specified SLA configuration
+   */
+  private boolean hasMissedSLA(long datasetLastRefreshTime, Map<String, Object> slaProps, long slaDetectionEndTime) {
+    // fetch the user configured SLA, otherwise default 3_DAYS.
+    long delay = TimeGranularity.fromString(MapUtils.getString(slaProps,  "sla", DEFAULT_DATA_SLA))
+        .toPeriod().toStandardDuration().getMillis();
+
+    return (slaDetectionEndTime - datasetLastRefreshTime) >= delay;
```

Comment:
We need to consider the expectedDelay in datasetconfig.
There are some datasets that have several days delay by default.

Reply:
The default dataset delay logic will apply when we automatically onboard all the alerts for SLA detection. The yaml translation logic needs to take care of it. 

Here, we expect that the sla is defined by the user and we will strictly adhere to it.